Client/Helpers/FxALoginHelper.swift

```diff
@@ -0,0 +1,191 @@
+
+import Account
+import Deferred
+import Foundation
+import Shared
+
+private let applicationDidRequestUserNotificationPermissionPrefKey = "applicationDidRequestUserNotificationPermissionPrefKey"
+
+/// This class manages the from successful login for FxAccounts to 
+/// asking the user for notification permissions, registering for 
+/// remote push notifications (APNS), registering for WebPush notifcations
+/// then creating an account and storing it in the profile.
+class FxALoginHelper {
+    private let deferred = Success()
+
+    private let pushClient: PushClient
+
+    private weak var application: UIApplication?
+    private weak var profile: Profile?
+
+    private var account: FirefoxAccount!
+    private var pushRegistration: PushRegistration!
+
+    static private(set) var sharedInstance: FxALoginHelper?
+
+    class func createSharedInstance(application: UIApplication?, profile: Profile?) -> FxALoginHelper {
+        sharedInstance = FxALoginHelper(application: application, profile: profile)
+        return sharedInstance!
+    }
+
+    init(application: UIApplication?, profile: Profile?) {
+        let configuration = DeveloperPushConfiguration()
+        let client = PushClient(endpointURL: configuration.endpointURL)
+
+        self.application = application
+        self.pushClient = client
+        self.profile = profile
+    }
+
+    func userDidLogin(data: JSON) -> Success {
+        if data["keyFetchToken"].asString == nil || data["unwrapBKey"].asString == nil {
+            // The /settings endpoint sends a partial "login"; ignore it entirely.
+            NSLog("Ignoring didSignIn with keyFetchToken or unwrapBKey missing.")
+            self.loginDidFail()
+            return deferred
+        }
+
+        // TODO: Error handling.
+        guard let profile = profile,
+            let account = FirefoxAccount.fromConfigurationAndJSON(profile.accountConfiguration, data: data) else {
+            self.loginDidFail()
+            return deferred
+        }
+
+        self.account = account
+        requestUserNotifications()
+        return deferred
+    }
+
+    func applicationDidLoadProfile() -> Success {
+        guard AppConstants.MOZ_FXA_PUSH else {
+            return finishNormally()
+        }
+
+        guard let account = profile?.getAccount() else {
+            // There's no account, no further action.
+            return finishNormally()
+        }
+
+        if let _ = account.pushRegistration {
+            // We have an account, and it's already registered for push notifications.
+            return finishNormally()
+        }
+
+        // Now: we have an account that does not have push notifications set up.
+        // however, we need to deal with cases of asking for permissions too frequently.
+        let asked = profile?.prefs.boolForKey(applicationDidRequestUserNotificationPermissionPrefKey) ?? true
+        let permitted = application!.currentUserNotificationSettings()!.types != .None
```

Comment:
This unwrap is worrisome because `application` is defined as a weak optional variable so we have no control over when this reference lives or dies. I would find a way to either wrap this with a guard or factor in the weakness of this property when considering the flow of the application at this point.